groups/bsl/bslstl/bslstl_ratio.t.cpp

```diff
@@ -0,0 +1,386 @@
+            ASSERT(            1000000ULL == bsl::mega::num);
+            ASSERT(         1000000000ULL == bsl::giga::num);
+            ASSERT(      1000000000000ULL == bsl::tera::num);
+            ASSERT(   1000000000000000ULL == bsl::peta::num);
+            ASSERT(1000000000000000000ULL == bsl::exa::num);
+
+        #ifdef BSL_RATIO_SUPPORTS_EXTENDED_SI_TYPEDEFS
+            ASSERT(1 == bsl::zepto::num);
+            ASSERT(1 == bsl::yocto::num);
+            ASSERT(   1000000000000000000000ULL == bsl::zepto::den);
+            ASSERT(1000000000000000000000000ULL == bsl::yocto::den);
+
+            ASSERT(1 == std::zetta::den);
+            ASSERT(1 == std::yotta::den);
+            ASSERT(   1000000000000000000000ULL == bsl::zetta::num);
+            ASSERT(1000000000000000000000000ULL == bsl::yotta::num);
+        #endif
+        }
+
+        if (verbose) printf("\tTest 'ratio' comparisons.\n");
+        {
+            typedef bsl::ratio<3, 2> ThreeHalfs;
+            typedef bsl::ratio<6, 4> Six4ths;
+            typedef bsl::ratio<0, 1> Zero;
+            typedef bsl::ratio<-2, 3> NegTwo3rds;
+
+            ASSERT( (bsl::ratio_less<NegTwo3rds, ThreeHalfs>::value));
+            ASSERT( (bsl::ratio_less<NegTwo3rds, Zero>::value));
+            ASSERT(!(bsl::ratio_less<ThreeHalfs, Zero>::value));
+            ASSERT(!(bsl::ratio_less<ThreeHalfs, Six4ths>::value));
+
+            ASSERT( (bsl::ratio_less_equal<NegTwo3rds, ThreeHalfs>::value));
+            ASSERT( (bsl::ratio_less_equal<NegTwo3rds, Zero>::value));
+            ASSERT(!(bsl::ratio_less_equal<ThreeHalfs, Zero>::value));
+            ASSERT( (bsl::ratio_less_equal<ThreeHalfs, Six4ths>::value));
+
+            ASSERT(!(bsl::ratio_greater<NegTwo3rds, ThreeHalfs>::value));
+            ASSERT(!(bsl::ratio_greater<NegTwo3rds, Zero>::value));
+            ASSERT( (bsl::ratio_greater<ThreeHalfs, Zero>::value));
+            ASSERT(!(bsl::ratio_greater<ThreeHalfs, Six4ths>::value));
+
+            ASSERT(!(bsl::ratio_greater_equal<NegTwo3rds, ThreeHalfs>::value));
+            ASSERT(!(bsl::ratio_greater_equal<NegTwo3rds, Zero>::value));
+            ASSERT( (bsl::ratio_greater_equal<ThreeHalfs, Zero>::value));
+            ASSERT( (bsl::ratio_greater_equal<ThreeHalfs, Six4ths>::value));
+
+            ASSERT(!(bsl::ratio_equal<NegTwo3rds, ThreeHalfs>::value));
+            ASSERT(!(bsl::ratio_equal<NegTwo3rds, Zero>::value));
+            ASSERT(!(bsl::ratio_equal<ThreeHalfs, Zero>::value));
+            ASSERT( (bsl::ratio_equal<ThreeHalfs, Six4ths>::value));
+
+            ASSERT( (bsl::ratio_not_equal<NegTwo3rds, ThreeHalfs>::value));
+            ASSERT( (bsl::ratio_not_equal<NegTwo3rds, Zero>::value));
+            ASSERT( (bsl::ratio_not_equal<ThreeHalfs, Zero>::value));
+            ASSERT(!(bsl::ratio_not_equal<ThreeHalfs, Six4ths>::value));
+
+#if defined(BSLS_LIBRARYFEATURES_HAS_CPP17_BASELINE_LIBRARY) || \
+    defined(BSLS_LIBRARYFEATURES_HAS_CPP14_BASELINE_LIBRARY)
```

Comment:
This is fine (unless Alisdair says otherwise).  But why I suggested:
```
#ifdef BSLS_LIBRARYFEATURES_HAS_CPP14_BASELINE_LIBRARY
```
is because if `BSLS_LIBRARYFEATURES_HAS_CPP17_BASELINE_LIBRARY` is defined then both `BSLS_LIBRARYFEATURES_HAS_CPP11_BASELINE_LIBRARY` and `BSLS_LIBRARYFEATURES_HAS_CPP14_BASELINE_LIBRARY` are defined.  Similarly, if `BSLS_LIBRARYFEATURES_HAS_CPP14_BASELINE_LIBRARY` is defined, then `BSLS_LIBRARYFEATURES_HAS_CPP11_BASELINE_LIBRARY` will be defined (per inspection of `bsls_libraryfeatures.h`). Note that your `ifndef` above takes advantage of what I just described.

Reply:
Ok. Thanks for the explanation